docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -0,0 +1,198 @@
+
+7. The node passes chunks sequentially to the app via `RequestApplySnapshotChunk`, along with the root chain hash at the snapshot height for verification. If the chunk cannot be verified or applied, the application returns `ResponseException` and Tendermint tries refetching the chunk. The final chunk has `final: true`.
+
+8. Once all chunks have been applied, the node compares the app hash to the chain hash, and if they do not match it either errors or discards the state and starts over.
+
+9. The node switches to fast sync to catch up blocks that were committed while restoring the snapshot.
+
+10. The node switches to normal consensus mode.
+
+## Gaia Proposal
+
+This describes the snapshot process seen from Gaia, using format version `1`. The serialization format is unspecified, but likely to be compressed Amino or Protobuf.
+
+### Snapshot Metadata
+
+In the initial version there is no snapshot metadata, so it is set to an empty byte buffer.
+
+Once all chunks have been successfully built, snapshot metadata should be serialized and stored in the file system as e.g. `snapshots/<height>/<format>/metadata`, and served via `RequestListSnapshots`.
+
+### Snapshot Chunk Format
+
+The Gaia data structure consists of a set of named IAVL trees. A root hash is constructed by taking the root hashes of each of the IAVL trees, then constructing a Merkle tree of the sorted name/hash map.
+
+IAVL trees are versioned, but a snapshot only contains the version relevant for the snapshot height. All historical versions are ignored.
+
+IAVL trees are insertion-order dependent, so key/value pairs must be stored in an appropriate insertion order to produce the same tree branching structure and thus the same Merkle hashes.
+
+A chunk corresponds to a subtree key range within an IAVL tree, in insertion order, along with the Merkle proof from the root of the subtree all the way up to the root of the multistore (including the Merkle proof for the surrounding multistore.
+
+```go
+struct SnapshotChunk {
+    Store   string            // Name (key) of IAVL store in outer MultiStore
+    Keys    [][]byte          // Snapshotted keys in insertion order
+    Values  [][]byte          // Snapshotted values corresponding to Keys
+    Proof   []merkle.ProofOp  // Merkle proof from subtree root to MultiStore root
+}
+```
+
+This chunk structure is believed to be sufficient to reconstruct an identical IAVL tree by applying separate chunks, but this may require the chunks to be ordered in a certain way; further research is needed.
+
+We do not use IAVL RangeProofs, since these include redundant data such as proofs for intermediate and leaf nodes that can be derived from the above data.
+
+Chunks should be built greedily by collecting key/value pairs constituting a complete subtree up to some size limit (e.g. 16 MB), then serialized and stored in the file system as `snapshots/<height>/<format>/<chunk>/data` and served via `RequestGetSnapshotChunk`.
```

Comment:
I would encourage you to checksum all the data you're going to store. E.g.:

```go
struct SnapshotChunk {
    ...
    Checksum [md5.Size]byte
}

This will allow you to detect broken snapshots before you send them to a peer requesting data.

Reply:
That seems like a good idea, I'll add a field.

Reply:
Actually, the checksum should be in the outer envelope, not the serialized payload, so that we can verify checksums without deserializing the payload. Have added this now.